pybind11/setup_helpers.py

```diff
@@ -0,0 +1,175 @@
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+
+import contextlib
+import os
+import shutil
+import sys
+import tempfile
+import threading
+
+import distutils.errors
+from distutils.command.build_ext import build_ext
+
+
+# It is recommended to use PEP 518 builds if using this module. However, this
+# file explicitly supports being copied into a user's project directory
+# standalone, and pulling pybind11 with the deprecated setup_requires feature.
+
+
+class DelayedPybindInclude(object):
+    """
+    Helper class to determine the pybind11 include path The purpose of this
+    class is to postpone importing pybind11 until it is actually installed, so
+    that the ``get_include()`` method can be invoked if pybind11 is loaded via
+    setup_requires.
+    """
+
+    def __str__(self):
+        import pybind11
+
+        return pybind11.get_include()
+
+
+# Just in case someone clever tries to multithread
+tmp_chdir_lock = threading.Lock()
+
+
+@contextlib.contextmanager
+def tmp_chdir():
+    "Prepare and enter a temporary directory, cleanup when done"
+    # Threadsafe
+
+    with tmp_chdir_lock:
+        olddir = os.getcwd()
+        try:
+            tmpdir = tempfile.mkdtemp()
+            os.chdir(tmpdir)
+            yield tmpdir
+        finally:
+            os.chdir(olddir)
+            shutil.rmtree(tmpdir)
+
+
+# cf http://bugs.python.org/issue26689
+def has_flag(compiler, flagname):
+    """
+    Return a boolean indicating whether a flag name is supported on the
+    specified compiler.
+    """
+
+    with tmp_chdir():
+        fname = "flagcheck.cpp"
+        with open(fname, "w") as f:
+            f.write("int main (int argc, char **argv) { return 0; }")
+
+        try:
+            compiler.compile([fname], extra_postargs=[flagname])
+            return True
+        except distutils.errors.CompileError:
+            return False
+
+
+def cpp_flag(compiler, value=None):
+    """
+    Return the ``-std=c++[11/14/17]`` compiler flag.
+    The newer version is preferred over c++11 (when it is available).
+    """
+
+    flags = ["-std=c++17", "-std=c++14", "-std=c++11"]
+
+    if value is not None:
+        mapping = {17: 0, 14: 1, 11: 2}
+        flags = [flags[mapping[value]]]
+
+    for flag in flags:
+        if has_flag(compiler, flag):
+            return flag
+
+    raise RuntimeError("Unsupported compiler -- at least C++11 support is needed!")
+
+
+c_opts = {
+    "msvc": ["/EHsc"],
+    "unix": [],
+}
+
+l_opts = {
+    "msvc": [],
+    "unix": [],
+}
+
+if sys.platform == "darwin":
+    darwin_opts = ["-stdlib=libc++", "-mmacosx-version-min=10.9"]
```

Comment:
I've got a version that checks for you in a separate branch. This initial one was just following the example provided in `pybind/`. I can probably merge those changes, though, since I now have tests working.

Reply:
Branch merged.